### PR TITLE
[WIP] Server-Side Apply: Alternative Status Wiping/Reset Fields implementation

### LINKED ISSUE #75564: server-side apply takes ownership of fields that don't get set due to status wiping
The API server wipes status changes on calls to the normal endpoints (not /status) at the storage registry layer (right next to etcd).  This means that calls to `<resource>/<name>` can't change status.  However, if you send a server-side apply patch, the server-side apply logic still takes ownership of any fields you *try* to set, despite the fact that they don't get set.

Reproduced on HEAD (8af1bf313efeeafb94c01ce04965ad8fe9cea21a).

Reproducer: 
- Send a server-side apply patch setting any field in status to `/api/v1/namespaces/<ns>/pods/<name>`
- Observe that ownership of that field has been taken
- Observe that the field isn't set

cc #73723 
cc @jennybuckley @apelisse

### PR DESCRIPTION
**Pull request by kwiesmueller**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

This is an alternative approach to https://github.com/kubernetes/kubernetes/pull/89486 which calls the respective preparation methods (status wiping, etc.) in the fieldManager before doing apply/update.

One disadvantage of this is, that there are additional conversions in the fieldManager to allow calling PrepareFor[Update|Create] on the internal objects.

Alternative implementation for the KEP https://github.com/kubernetes/enhancements/pull/1123.

**Which issue(s) this PR fixes**:

Fixes #75564

**Does this PR introduce a user-facing change?**:

```release-note
FieldManager no longer owns fields that get reset before the object is persisted (e.g. "status wiping").
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/1123
```

/sig api-machinery
/wg api-expression
/cc @apelisse @jennybuckley @julianvmodesto

### CODE REVIEW

**Comment by kwiesmueller** [comment]
There is a failure in the tests which is not in the other approach:
```
resetFields_test.go:118: CertificateSigningRequest.certificates.k8s.io "csr1" is invalid: [spec.request: Invalid value: []byte(nil): PEM block type must be CERTIFICATE REQUEST, spec.usages: Required value: usages must be provided, spec.signerName: Required value: signerName must be provided]
```

This could be due to the conversion. I'll investigate.

Also this approach still requires a few/many changes to our internal resources, because not all of them implement the GenericStore and neither do their subresources.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/structuredmerge.go [hunk 1]
```diff
@@ -83,16 +87,31 @@ func NewCRDStructuredMergeManager(models openapiproto.Models, objectConverter ru
...
 }
 
 // Update implements Manager.
 func (f *structuredMergeManager) Update(liveObj, newObj runtime.Object, managed Managed, manager string) (runtime.Object, Managed, error) {
-	newObjVersioned, err := f.toVersioned(newObj)
+	newObjUnversioned, err := f.toUnversioned(newObj)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to convert new object to unversioned: %v", err)
+	}
+	liveObjUnversioned, err := f.toUnversioned(liveObj)
```
Comment (kwiesmueller) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/structuredmerge.go:100]: @apelisse it seems like i still have to do the conversion roundtrip for both objects as we don't get unversioned.
The reason we do .toVersioned below seems to be ensuring the right version?

**Comment by kwiesmueller** [comment]
And for reference, the current list of resource paths not implementing the required interfaces (like GenericStore):

```
apiextensions.k8s.io v1 customresourcedefinitions/status
apiextensions.k8s.io v1beta1 customresourcedefinitions/status
core v1 bindings
core v1 componentstatuses
core v1 namespaces
core v1 namespaces/finalize
core v1 namespaces/status
core v1 nodes/proxy
core v1 nodes/status
core v1 persistentvolumeclaims/status
core v1 persistentvolumes/status
core v1 pods/attach
core v1 pods/binding
core v1 pods/eviction
core v1 pods/exec
core v1 pods/log
core v1 pods/portforward
core v1 pods/proxy
core v1 pods/status
core v1 replicationcontrollers/scale
core v1 replicationcontrollers/status
core v1 resourcequotas/status
core v1 services
core v1 services/proxy
core v1 services/status
authentication.k8s.io v1 tokenreviews
authentication.k8s.io v1beta1 tokenreviews
authorization.k8s.io v1 localsubjectaccessreviews
authorization.k8s.io v1 selfsubjectaccessreviews
authorization.k8s.io v1 selfsubjectrulesreviews
authorization.k8s.io v1 subjectaccessreviews
authorization.k8s.io v1beta1 localsubjectaccessreviews
authorization.k8s.io v1beta1 selfsubjectaccessreviews
authorization.k8s.io v1beta1 selfsubjectrulesreviews
authorization.k8s.io v1beta1 subjectaccessreviews
autoscaling v1 horizontalpodautoscalers/status
autoscaling v2beta1 horizontalpodautoscalers/status
autoscaling v2beta2 horizontalpodautoscalers/status
batch v1 jobs/status
batch v1beta1 cronjobs/status
batch v2alpha1 cronjobs/status
certificates.k8s.io v1beta1 certificatesigningrequests/approval
certificates.k8s.io v1beta1 certificatesigningrequests/status
extensions v1beta1 ingresses/status
networking.k8s.io v1beta1 ingresses/status
policy v1beta1 poddisruptionbudgets/status
rbac.authorization.k8s.io v1 clusterrolebindings
rbac.authorization.k8s.io v1 clusterroles
rbac.authorization.k8s.io v1 rolebindings
rbac.authorization.k8s.io v1 roles
rbac.authorization.k8s.io v1beta1 clusterrolebindings
rbac.authorization.k8s.io v1beta1 clusterroles
rbac.authorization.k8s.io v1beta1 rolebindings
rbac.authorization.k8s.io v1beta1 roles
rbac.authorization.k8s.io v1alpha1 clusterrolebindings
rbac.authorization.k8s.io v1alpha1 clusterroles
rbac.authorization.k8s.io v1alpha1 rolebindings
rbac.authorization.k8s.io v1alpha1 roles
storage.k8s.io v1 volumeattachments/status
flowcontrol.apiserver.k8s.io v1alpha1 flowschemas/status
flowcontrol.apiserver.k8s.io v1alpha1 prioritylevelconfigurations/status
apps v1 daemonsets/status
apps v1 deployments/scale
apps v1 deployments/status
apps v1 replicasets/scale
apps v1 replicasets/status
apps v1 statefulsets/scale
apps v1 statefulsets/status
```

Quite a few, but then there is also ones that are not in scope right now.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 2]
```diff
@@ -867,6 +866,18 @@ func (r *crdHandler) getOrCreateServingInfoFor(uid types.UID, name string) (*crd
 			SelfLinkPathPrefix: selfLinkPrefix,
 			SelfLinkPathSuffix: "/status",
 		}
+		if utilfeature.DefaultFeatureGate.Enabled(features.ServerSideApply) {
```
Comment (apelisse) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:869]: Did we forget to create the fieldmanager before?

Comment (apelisse) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:869]: I'm pretty sure we have a test where we do status update on CR that has the managed fields?

Comment (kwiesmueller) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:869]: No, previously it just took the existing manager from the parent resource.
You can see how we overwrite it for scale to nil. But it was there.
This just creates a new one to pass a different strategy

**Pull request closed without merging**